crates/sui-core/src/unit_tests/move_package_upgrade_tests.rs

```diff
@@ -990,3 +1002,115 @@ async fn test_publish_transitive_override_happy_path() {
+
+    // First upgrade (version 2) introduces a new type, B.
+    let (digest, modules) = build_upgrade_test_modules("makes_new_object");
+    let effects = runner
+        .upgrade(
+            UpgradePolicy::COMPATIBLE,
+            digest,
+            modules,
+            vec![SUI_FRAMEWORK_OBJECT_ID, MOVE_STDLIB_OBJECT_ID],
+        )
+        .await;
+
+    assert!(effects.status.is_ok(), "{:#?}", effects.status);
+    let package_v2 = runner.package.0;
+
+    // Second upgrade (version 3) introduces a new type, C.
+    let (digest, modules) = build_upgrade_test_modules("makes_another_object");
+    let effects = runner
+        .upgrade(
+            UpgradePolicy::COMPATIBLE,
+            digest,
+            modules,
+            vec![SUI_FRAMEWORK_OBJECT_ID, MOVE_STDLIB_OBJECT_ID],
+        )
+        .await;
+
+    assert!(effects.status.is_ok(), "{:#?}", effects.status);
+    let package_v3 = runner.package.0;
+
+    // Create an instance of the type introduced at version 2 using function from version 2.
+    let TransactionEffects::V1(effects) = runner
+        .run({
+            let mut builder = ProgrammableTransactionBuilder::new();
+            move_call! { builder, (package_v2)::base::makes_b() };
+            builder.finish()
+        })
+        .await;
+
+    assert!(effects.status.is_ok(), "{:#?}", effects.status);
+    let created_b = effects
+        .created
+        .iter()
+        .find_map(|(b, owner)| matches!(owner, Owner::AddressOwner(_)).then_some(b))
+        .unwrap();
+
+    // Create an instance of the type introduced at version 3 using function from version 3.
+    let TransactionEffects::V1(effects) = runner
+        .run({
+            let mut builder = ProgrammableTransactionBuilder::new();
+            move_call! { builder, (package_v3)::base::makes_c() };
+            builder.finish()
+        })
+        .await;
+
+    assert!(effects.status.is_ok(), "{:#?}", effects.status);
+    let created_c = effects
+        .created
+        .iter()
+        .find_map(|(c, owner)| matches!(owner, Owner::AddressOwner(_)).then_some(c))
+        .unwrap();
+
+    // modify objects created of types introduced at versions 2 and 3 and emit events using types
+    // introduced at versions 2 and 3 (using functions from version 3)
+    let TransactionEffects::V1(effects) = runner
+        .run({
+            let mut builder = ProgrammableTransactionBuilder::new();
+            let b = builder
+                .obj(ObjectArg::ImmOrOwnedObject(*created_b))
+                .unwrap();
+            move_call! { builder, (package_v3)::base::modifies_b(b) };
+            let c = builder
+                .obj(ObjectArg::ImmOrOwnedObject(*created_c))
+                .unwrap();
+            move_call! { builder, (package_v3)::base::modifies_c(c) };
+            builder.finish()
+        })
+        .await;
+
+    assert!(effects.status.is_ok(), "{:#?}", effects.status);
```

Comment:
Can you add assertions about the expected events (i.e. the `BModEvent` and `CModEvent`) that get produced as well? (Fine to do as a follow-up).

Reply:
Done!